lib/modules/manager/pip-compile/extract.ts

```diff
@@ -216,19 +220,29 @@ export async function extractAllPackageFiles(
   // This needs to go in reverse order to handle transitive dependencies
   for (const packageFile of [...result].reverse()) {
     for (const reqFile of packageFile.managerData?.requirementsFiles ?? []) {
-      let sourceFile: PackageFile | undefined = undefined;
+      let sourceFiles: PackageFile[] | undefined = undefined;
       if (fileMatches.includes(reqFile)) {
-        sourceFile = lockFileSources.get(reqFile);
+        sourceFiles = lockFileSources.get(reqFile);
       } else if (packageFiles.has(reqFile)) {
-        sourceFile = packageFiles.get(reqFile);
+        sourceFiles = [packageFiles.get(reqFile)!];
       }
-      if (!sourceFile) {
+      if (!sourceFiles) {
         logger.warn(
           `pip-compile: ${packageFile.packageFile} references ${reqFile} which does not appear to be a requirements file managed by pip-compile`,
         );
         continue;
       }
-      sourceFile.lockFiles!.push(...packageFile.lockFiles!);
+      const files = new Set([...packageFile.lockFiles!].reverse());
```

Comment:
```suggestion
      // These get reversed before merging so that we keep the last instance of any common
      // lock files, since a file that -r includes multiple lock files needs to be updated after
      // all of the lock files it includes
      const files = new Set([...packageFile.lockFiles!].reverse());
```